### PR TITLE
Refactor init for Excel readers to _BaseExcelReader

### PR DESCRIPTION
**Pull request by tdamsma**
Pre-cursor for #25092. To support multiple excel readers the init function is moved to _BaseExcelReader. The function is made a bit more general to be able to support multiple excel readers.

- [ ] tests passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
looks good. ping when ready. cc @WillAyd

File: pandas/io/excel/_xlrd.py [hunk 1]
```diff
@@ -30,35 +28,20 @@ def __init__(self, filepath_or_buffer):
...
-
-        if isinstance(filepath_or_buffer, xlrd.Book):
-            self.book = filepath_or_buffer
-        elif hasattr(filepath_or_buffer, "read"):
-            # N.B. xlrd.Book has a read attribute too
-            if hasattr(filepath_or_buffer, 'seek'):
-                try:
-                    # GH 19779
-                    filepath_or_buffer.seek(0)
-                except UnsupportedOperation:
```
Comment (WillAyd) [inline pandas/io/excel/_xlrd.py]: I see this wasn't carried over - is it no longer valid?

File: pandas/io/excel/_xlrd.py [hunk 2]
```diff
@@ -30,35 +28,20 @@ def __init__(self, filepath_or_buffer):
...
-                    # HTTPResponse does not support seek()
-                    # GH 20434
-                    pass
+        self._engine = xlrd
+        super(_XlrdReader, self).__init__(filepath_or_buffer)
```
Comment (WillAyd) [inline pandas/io/excel/_xlrd.py]: We just switched to 3.0 syntax so just call `super()` here

File: pandas/io/excel/_xlrd.py [hunk 3]
```diff
@@ -30,35 +28,20 @@ def __init__(self, filepath_or_buffer):
...
+
+    @property
+    def _workbook_class(self):
+        return self._engine.Book
```
Comment (WillAyd) [inline pandas/io/excel/_xlrd.py]: I think it would be better if this was namespaced directly from xlrd instead of going through `self._engine`

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/excel/_xlrd.py (hunk 1 above)
Comment (tdamsma) [reply pandas/io/excel/_xlrd.py]: This is now dealt with by wrapping url responses in `BytesIO(urlopen(filepath_or_buffer).read())`. So now all items with `"read"` can also support `seek(0)`

File: pandas/io/excel/_xlrd.py (hunk 2 above)
Comment (tdamsma) [reply pandas/io/excel/_xlrd.py]: agreed, implemented

File: pandas/io/excel/_xlrd.py (hunk 3 above)
Comment (tdamsma) [reply pandas/io/excel/_xlrd.py]: agreed, implemented

**Comment by tdamsma** [comment]
@jreback, @WillAyd, I don't understand the test failing tests, but other than that I think it's ready

**Comment by WillAyd** [comment]
@tdamsma restarted all of them. Not sure what is wrong with the checks_and_doc job off the top of my head but looks like this PR introduced some regressions with handling of S3 excel files as shown in the Travis failures. 

If you could take a look would be great

**Comment by jreback** [comment]
lgtm. @WillAyd

**Comment by tdamsma** [comment]
@WillAyd, I addressed the regression. Also I accidentally committed (and reverted) a setting to limit the test scope, that explains the codecov report.

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Nice job @tdamsma - looking forward to the follow ups!

**Comment by tdamsma** [comment]
Great, thanks for the reviews

**Pull request merged**